### ISSUE TITLE
Firefox 120 crashing on WebGL on NixOS 23.05

### ISSUE THREAD

**Issue by sigprof** [NixOS/nixpkgs#269571]
### Describe the bug
After the recent update of nixos-23.05 from f741f8a839912e272d7e87ccf4b9dbc6012cdaf9 to 8f1180704ac35baded1a74164365ac7cdfba6f38 (which included https://github.com/NixOS/nixpkgs/commit/197b65e46dafcb7167964bf4af6700ab7abc52a0, updating Firefox from 119.0.1 to 120.0) Firefox started crashing, apparently when WebGL gets used.

### Steps To Reproduce
1. Start `firefox` 120.0 with a clean profile.
2. Open https://webglsamples.org/aquarium/aquarium.html
3. Observe a crash.

### Additional context
Some crash reports:
- https://crash-stats.mozilla.org/report/index/bp-36fb9213-b16b-444e-8689-5b4620231124
- https://crash-stats.mozilla.org/report/index/bp-9e0300b5-8850-4638-9974-1dc1f0231124

These crash reports point to https://bugzilla.mozilla.org/show_bug.cgi?id=1861365, which shows a matching backtrace:
```
0	firefox	std::locale::operator=(std::locale const&)		context
1	firefox	std::ios_base::imbue(std::locale const&)		cfi
2	firefox	std::basic_ios<char, std::char_traits<char> >::imbue(std::locale const&)		cfi
3	libxul.so	sh::InitializeStream<std::__cxx11::basic_ostringstream<char, std::char_traits<char>, std::allocator<char> > >()	/build/firefox-120.0/gfx/angle/checkout/src/compiler/translator/Common.h:238	cfi
4	libxul.so	sh::TCompiler::setResourceString()	/build/firefox-120.0/gfx/angle/checkout/src/compiler/translator/Compiler.cpp:1294	inlined
4	libxul.so	sh::TCompiler::Init(ShBuiltInResources const&)	/build/firefox-120.0/gfx/angle/checkout/src/compiler/translator/Compiler.cpp:407	cfi
5	libxul.so	sh::ConstructCompiler(unsigned int, ShShaderSpec, ShShaderOutput, ShBuiltInResources const*)	/build/firefox-120.0/gfx/angle/checkout/src/compiler/translator/ShaderLang.cpp:368	cfi
6	libxul.so	mozilla::webgl::ShaderValidator::Create(unsigned int, ShShaderSpec, ShShaderOutput, ShBuiltInResources const&, ShCompileOptions)	/build/firefox-120.0/dom/canvas/WebGLShaderValidator.cpp:215	inlined
6	libxul.so	mozilla::WebGLContext::CreateShaderValidator(unsigned int) const	/build/firefox-120.0/dom/canvas/WebGLShaderValidator.cpp:196	inlined
6	libxul.so	mozilla::WebGLShader::CompileShader()	/build/firefox-120.0/dom/canvas/WebGLShader.cpp:98	cfi
```
The underlying problem might be somewhat different though, because the offending `-static-libstdc++` option is not present in any Hydra build logs (`-Bsymbolic-functions` is present though).

### Notify maintainers

@lovesegfault @mweinelt 

### Metadata
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.63, NixOS, 23.05 (Stoat), 23.05.20231122.8f11807`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.13.6`
 - nixpkgs: `/etc/nix/channels/nixpkgs`

**Comment by BadCoder12** [comment]
You can disable WebGL temporarily while the bug is being fixed.
about:config
webgl.disabled set to true

**Comment by sigprof** [comment]
`firefox-bin` also works without crashing even with WebGL enabled.

**Comment by sigprof** [comment]
Also apparently the problem is specific to the `nixos-23.05` branch:
```
❯ nm -D $(nix path-info github:NixOS/nixpkgs/nixos-23.05#firefox-unwrapped)/bin/.firefox-wrapped | c++filt | grep ' T std::locale'
00000000000c33c0 T std::locale::id::_M_id() const
00000000000c2950 T std::locale::name[abi:cxx11]() const
00000000000d0510 T std::locale::name() const
00000000000d4920 T std::locale::facet::_M_cow_shim(std::locale::id const*) const
00000000000f28e0 T std::locale::facet::_M_sso_shim(std::locale::id const*) const
00000000000c2ca0 T std::locale::operator==(std::locale const&) const
00000000000c55c0 T std::locale::_S_initialize()
00000000000c5560 T std::locale::_S_initialize_once()
00000000000c2df0 T std::locale::_S_normalize_category(int)
00000000000c2e60 T std::locale::facet::_S_get_c_name()
00000000000c2e30 T std::locale::facet::_S_get_c_locale()
00000000000c3ff0 T std::locale::facet::_S_clone_c_locale(__locale_struct*&)
00000000000c3f80 T std::locale::facet::_S_create_c_locale(__locale_struct*&, char const*, __locale_struct*)
00000000000c2860 T std::locale::facet::_S_initialize_once()
00000000000c3fb0 T std::locale::facet::_S_destroy_c_locale(__locale_struct*&)
00000000000c4000 T std::locale::facet::_S_lc_ctype_c_locale(__locale_struct*, char const*)
00000000000c2830 T std::locale::facet::~facet()
00000000000c2810 T std::locale::facet::~facet()
00000000000c2810 T std::locale::facet::~facet()
00000000000cf5a0 T std::locale::_Impl::_M_init_extra(std::locale::facet**)
00000000000cfdf0 T std::locale::_Impl::_M_init_extra(void*, void*, char const*, char const*)
00000000000c3b20 T std::locale::_Impl::_M_install_cache(std::locale::facet const*, unsigned long)
00000000000c3510 T std::locale::_Impl::_M_install_facet(std::locale::id const*, std::locale::facet const*)
00000000000c3a00 T std::locale::_Impl::_M_replace_facet(std::locale::_Impl const*, std::locale::id const*)
00000000000c3a50 T std::locale::_Impl::_M_replace_category(std::locale::_Impl const*, std::locale::id const* const*)
00000000000c4750 T std::locale::_Impl::_Impl(unsigned long)
00000000000c31f0 T std::locale::_Impl::_Impl(std::locale::_Impl const&, unsigned long)
00000000000c4750 T std::locale::_Impl::_Impl(unsigned long)
00000000000c31f0 T std::locale::_Impl::_Impl(std::locale::_Impl const&, unsigned long)
00000000000c2ed0 T std::locale::_Impl::~_Impl()
00000000000c2ed0 T std::locale::_Impl::~_Impl()
00000000000c57f0 T std::locale::global(std::locale const&)
00000000000c5770 T std::locale::classic()
00000000000c30c0 T std::locale::operator=(std::locale const&)
00000000000c2930 T std::locale::locale(std::locale::_Impl*)
00000000000c28f0 T std::locale::locale(std::locale const&)
00000000000c5690 T std::locale::locale()
00000000000c2930 T std::locale::locale(std::locale::_Impl*)
00000000000c28f0 T std::locale::locale(std::locale const&)
00000000000c5690 T std::locale::locale()
00000000000c3170 T std::locale::~locale()
00000000000c3170 T std::locale::~locale()

❯ nm -D $(nix path-info github:NixOS/nixpkgs/nixos-unstable#firefox-unwrapped)/bin/.firefox-wrapped | c++filt | grep ' T std::locale' | wc -l
0
```

**Comment by sigprof** [comment]
Comparing the “good” build log (https://hydra.nixos.org/build/241859143) with the “bad” one (https://hydra.nixos.org/build/242110774) shows the difference in clang versions (and the extra `-std=gnu++17` option for the `nixos-23.05` build, which uses clang-15):
```diff
-/nix/store/0szfvpknzs1ykh5s3zaavskq42kp7qvv-clang-wrapper-15.0.7/bin/clang++ -std=gnu++17
+/nix/store/cpaw2fx5vdf7f2m552klp8ldasfn5mga-clang-wrapper-16.0.6/bin/clang++
```
And the difference is indeed in the compiler behavior.  Using the `pain.cc` test file from the Mozilla bug report:
```c++
#include <iostream>
void pain() {
  std::cout << "[pain] " << std::locale::classic().name() <<"\n";
}
```
With the new compiler from `nixos-unstable`:
```
❯ /nix/store/cpaw2fx5vdf7f2m552klp8ldasfn5mga-clang-wrapper-16.0.6/bin/clang++ -v -fPIC -shared -o libpain.so pain.cc
clang version 16.0.6
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /nix/store/5ql31bfsgynanf794wnl2wlbs23q7b8h-clang-16.0.6/bin
Found candidate GCC installation: /nix/store/hf2gy3km07d5m0p1lwmja0rg9wlnmyr7-gcc-12.3.0/lib/gcc/x86_64-unknown-linux-gnu/12.3.0
Found candidate GCC installation: /nix/store/hf2gy3km07d5m0p1lwmja0rg9wlnmyr7-gcc-12.3.0/lib64/gcc/x86_64-unknown-linux-gnu/12.3.0
Selected GCC installation: /nix/store/hf2gy3km07d5m0p1lwmja0rg9wlnmyr7-gcc-12.3.0/lib64/gcc/x86_64-unknown-linux-gnu/12.3.0
Candidate multilib: .;@m64
Selected multilib: .;@m64
 "/nix/store/5ql31bfsgynanf794wnl2wlbs23q7b8h-clang-16.0.6/bin/clang-16" -cc1 -triple x86_64-unknown-linux-gnu -emit-obj -mrelax-all -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name pain.cc -mrelocation-model pic -pic-level 2 -fhalf-no-semantic-interposition -mframe-pointer=all -fmath-errno -ffp-contract=on -fno-rounding-math -mconstructor-aliases -funwind-tables=2 -target-cpu x86-64 -tune-cpu generic -mllvm -treat-scalable-fixed-error-as-warning -debugger-tuning=gdb -v -fcoverage-compilation-dir=/home/vsu/Projects/Junk/libstdcxx-test -nostdsysteminc -resource-dir /nix/store/cpaw2fx5vdf7f2m552klp8ldasfn5mga-clang-wrapper-16.0.6/resource-root -idirafter /nix/store/mrgib0s2ayr81xv1q84xsjg8ijybalq3-glibc-2.38-27-dev/include -isystem /nix/store/hf2gy3km07d5m0p1lwmja0rg9wlnmyr7-gcc-12.3.0/include/c++/12.3.0 -isystem /nix/store/hf2gy3km07d5m0p1lwmja0rg9wlnmyr7-gcc-12.3.0/include/c++/12.3.0/x86_64-unknown-linux-gnu -internal-isystem /nix/store/cpaw2fx5vdf7f2m552klp8ldasfn5mga-clang-wrapper-16.0.6/resource-root/include -fdeprecated-macro -fdebug-compilation-dir=/home/vsu/Projects/Junk/libstdcxx-test -ferror-limit 19 -fgnuc-version=4.2.1 -fcxx-exceptions -fexceptions -fcolor-diagnostics -faddrsig -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o /tmp/pain-3ea9df.o -x c++ pain.cc
clang -cc1 version 16.0.6 based upon LLVM 16.0.6 default target x86_64-unknown-linux-gnu
#include "..." search starts here:
#include <...> search starts here:
 /nix/store/hf2gy3km07d5m0p1lwmja0rg9wlnmyr7-gcc-12.3.0/include/c++/12.3.0
 /nix/store/hf2gy3km07d5m0p1lwmja0rg9wlnmyr7-gcc-12.3.0/include/c++/12.3.0/x86_64-unknown-linux-gnu
 /nix/store/cpaw2fx5vdf7f2m552klp8ldasfn5mga-clang-wrapper-16.0.6/resource-root/include
 /nix/store/mrgib0s2ayr81xv1q84xsjg8ijybalq3-glibc-2.38-27-dev/include
End of search list.
 "/nix/store/cpaw2fx5vdf7f2m552klp8ldasfn5mga-clang-wrapper-16.0.6/bin/ld" --hash-style=gnu --eh-frame-hdr -m elf_x86_64 -shared -o libpain.so /nix/store/qn3ggz5sf3hkjs2c797xf7nan3amdxmp-glibc-2.38-27/lib/crti.o /nix/store/hf2gy3km07d5m0p1lwmja0rg9wlnmyr7-gcc-12.3.0/lib/gcc/x86_64-unknown-linux-gnu/12.3.0/crtbeginS.o -L/nix/store/qn3ggz5sf3hkjs2c797xf7nan3amdxmp-glibc-2.38-27/lib -L/nix/store/hf2gy3km07d5m0p1lwmja0rg9wlnmyr7-gcc-12.3.0/lib/gcc/x86_64-unknown-linux-gnu/12.3.0 -L/nix/store/myw67gkgayf3s2mniij7zwd79lxy8v0k-gcc-12.3.0-lib/lib -L/nix/store/hf2gy3km07d5m0p1lwmja0rg9wlnmyr7-gcc-12.3.0//lib -L/nix/store/7vq9zymw7zk2i0k1i4nfkl6hmlx3gxji-gcc-12.3.0-libgcc/lib -L/nix/store/d853i1aciijvq2bp75nb22vr04qp3v24-clang-16.0.6-lib/lib -L/nix/store/hf2gy3km07d5m0p1lwmja0rg9wlnmyr7-gcc-12.3.0/lib64/gcc/x86_64-unknown-linux-gnu/12.3.0 -L/nix/store/hf2gy3km07d5m0p1lwmja0rg9wlnmyr7-gcc-12.3.0/lib64/gcc/x86_64-unknown-linux-gnu/12.3.0/../../../../lib64 -dynamic-linker=/nix/store/qn3ggz5sf3hkjs2c797xf7nan3amdxmp-glibc-2.38-27/lib/ld-linux-x86-64.so.2 /tmp/pain-3ea9df.o -lstdc++ -lm -lgcc_s -lgcc -lc -lgcc_s -lgcc /nix/store/hf2gy3km07d5m0p1lwmja0rg9wlnmyr7-gcc-12.3.0/lib/gcc/x86_64-unknown-linux-gnu/12.3.0/crtendS.o /nix/store/qn3ggz5sf3hkjs2c797xf7nan3amdxmp-glibc-2.38-27/lib/crtn.o

❯ /nix/store/cpaw2fx5vdf7f2m552klp8ldasfn5mga-clang-wrapper-16.0.6/bin/nm -D libpain.so | grep ' T '
0000000000001b20 T _Z4painv
```

With the old compiler from `nixos-23.05`:
```
❯ /nix/store/0szfvpknzs1ykh5s3zaavskq42kp7qvv-clang-wrapper-15.0.7/bin/clang++ -v -fPIC -shared -o libpain.so pain.cc
clang version 15.0.7
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /nix/store/c0mn6zjclhqx9l2y1ny088kphcgkk84i-clang-15.0.7/bin
Found candidate GCC installation: /nix/store/jgdz5wfqc6q6qhy17lsqnfpbkfi1fcar-gcc-12.2.0/lib/gcc/x86_64-unknown-linux-gnu/12.2.0
Found candidate GCC installation: /nix/store/jgdz5wfqc6q6qhy17lsqnfpbkfi1fcar-gcc-12.2.0/lib64/gcc/x86_64-unknown-linux-gnu/12.2.0
Selected GCC installation: /nix/store/jgdz5wfqc6q6qhy17lsqnfpbkfi1fcar-gcc-12.2.0/lib64/gcc/x86_64-unknown-linux-gnu/12.2.0
Candidate multilib: .;@m64
Selected multilib: .;@m64
 "/nix/store/c0mn6zjclhqx9l2y1ny088kphcgkk84i-clang-15.0.7/bin/clang-15" -cc1 -triple x86_64-unknown-linux-gnu -emit-obj -mrelax-all --mrelax-relocations -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name pain.cc -mrelocation-model pic -pic-level 2 -fhalf-no-semantic-interposition -mframe-pointer=all -fmath-errno -ffp-contract=on -fno-rounding-math -mconstructor-aliases -funwind-tables=2 -target-cpu x86-64 -tune-cpu generic -mllvm -treat-scalable-fixed-error-as-warning -debugger-tuning=gdb -v -fcoverage-compilation-dir=/home/vsu/Projects/Junk/libstdcxx-test -nostdsysteminc -resource-dir /nix/store/0szfvpknzs1ykh5s3zaavskq42kp7qvv-clang-wrapper-15.0.7/resource-root -idirafter /nix/store/linb70pgp53f0kygxm5wk163dx22d681-glibc-2.37-45-dev/include -isystem /nix/store/jgdz5wfqc6q6qhy17lsqnfpbkfi1fcar-gcc-12.2.0/include/c++/12.2.0 -isystem /nix/store/jgdz5wfqc6q6qhy17lsqnfpbkfi1fcar-gcc-12.2.0/include/c++/12.2.0/x86_64-unknown-linux-gnu -internal-isystem /nix/store/0szfvpknzs1ykh5s3zaavskq42kp7qvv-clang-wrapper-15.0.7/resource-root/include -fdeprecated-macro -fdebug-compilation-dir=/home/vsu/Projects/Junk/libstdcxx-test -ferror-limit 19 -fgnuc-version=4.2.1 -fcxx-exceptions -fexceptions -fcolor-diagnostics -faddrsig -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o /tmp/pain-d328d9.o -x c++ pain.cc
clang -cc1 version 15.0.7 based upon LLVM 15.0.7 default target x86_64-unknown-linux-gnu
#include "..." search starts here:
#include <...> search starts here:
 /nix/store/jgdz5wfqc6q6qhy17lsqnfpbkfi1fcar-gcc-12.2.0/include/c++/12.2.0
 /nix/store/jgdz5wfqc6q6qhy17lsqnfpbkfi1fcar-gcc-12.2.0/include/c++/12.2.0/x86_64-unknown-linux-gnu
 /nix/store/0szfvpknzs1ykh5s3zaavskq42kp7qvv-clang-wrapper-15.0.7/resource-root/include
 /nix/store/linb70pgp53f0kygxm5wk163dx22d681-glibc-2.37-45-dev/include
End of search list.
 "/nix/store/0szfvpknzs1ykh5s3zaavskq42kp7qvv-clang-wrapper-15.0.7/bin/ld" --eh-frame-hdr -m elf_x86_64 -shared -o libpain.so /nix/store/whypqfa83z4bsn43n4byvmw80n4mg3r8-glibc-2.37-45/lib/crti.o /nix/store/jgdz5wfqc6q6qhy17lsqnfpbkfi1fcar-gcc-12.2.0/lib/gcc/x86_64-unknown-linux-gnu/12.2.0/crtbeginS.o -L/nix/store/whypqfa83z4bsn43n4byvmw80n4mg3r8-glibc-2.37-45/lib -L/nix/store/jgdz5wfqc6q6qhy17lsqnfpbkfi1fcar-gcc-12.2.0/lib/gcc/x86_64-unknown-linux-gnu/12.2.0 -L/nix/store/bmfc7mrr7kilf45wpfd60davjaqfmagv-gcc-12.2.0-lib/x86_64-unknown-linux-gnu/lib -L/nix/store/jgdz5wfqc6q6qhy17lsqnfpbkfi1fcar-gcc-12.2.0//lib -L/nix/store/8vfh3zzdrsrg8v2763n4acy622ai2h2c-gcc-12.2.0-libgcc/lib -L/nix/store/irg3aj4ikr12kszdvk83m5q544wjs3sh-clang-15.0.7-lib/lib -L/nix/store/jgdz5wfqc6q6qhy17lsqnfpbkfi1fcar-gcc-12.2.0/lib64/gcc/x86_64-unknown-linux-gnu/12.2.0 -L/nix/store/jgdz5wfqc6q6qhy17lsqnfpbkfi1fcar-gcc-12.2.0/lib64/gcc/x86_64-unknown-linux-gnu/12.2.0/../../../../lib64 -dynamic-linker=/nix/store/whypqfa83z4bsn43n4byvmw80n4mg3r8-glibc-2.37-45/lib/ld-linux-x86-64.so.2 /tmp/pain-d328d9.o -lstdc++ -lm -lgcc_s -lgcc -lc -lgcc_s -lgcc /nix/store/jgdz5wfqc6q6qhy17lsqnfpbkfi1fcar-gcc-12.2.0/lib/gcc/x86_64-unknown-linux-gnu/12.2.0/crtendS.o /nix/store/whypqfa83z4bsn43n4byvmw80n4mg3r8-glibc-2.37-45/lib/crtn.o

❯ /nix/store/cpaw2fx5vdf7f2m552klp8ldasfn5mga-clang-wrapper-16.0.6/bin/nm -D libpain.so | grep ' T ' | grep _S_initialize
00000000000bbc80 T _ZNSt6locale13_S_initializeEv
00000000000bbc20 T _ZNSt6locale18_S_initialize_onceEv
00000000000b8c50 T _ZNSt6locale5facet18_S_initialize_onceEv

❯ /nix/store/cpaw2fx5vdf7f2m552klp8ldasfn5mga-clang-wrapper-16.0.6/bin/nm -D libpain.so | grep ' T ' | wc -l
768
```

Apparently the important difference here is that the compiler from `nixos-unstable` passes the `-L/nix/store/myw67gkgayf3s2mniij7zwd79lxy8v0k-gcc-12.3.0-lib/lib` option to `ld`, and the `/nix/store/myw67gkgayf3s2mniij7zwd79lxy8v0k-gcc-12.3.0-lib/lib` directory contains `libstdc++.so`, therefore the linker uses the definitions from the `libstdc++.so` shared library as intended.  However, the compiler from `nixos-23.05` passes the `-L/nix/store/bmfc7mrr7kilf45wpfd60davjaqfmagv-gcc-12.2.0-lib/x86_64-unknown-linux-gnu/lib` option, which refers to a nonexistent directory (`/nix/store/bmfc7mrr7kilf45wpfd60davjaqfmagv-gcc-12.2.0-lib/lib` actually exists, and there is the `/nix/store/bmfc7mrr7kilf45wpfd60davjaqfmagv-gcc-12.2.0-lib/lib64` symlink, but these directories are not mentioned in the `ld` parameters), therefore the linker does not see `libstdc++.so` and uses only `libstdc++.a` from ` -L/nix/store/jgdz5wfqc6q6qhy17lsqnfpbkfi1fcar-gcc-12.2.0//lib`, so the resulting shared library contains copies of all libstdc++ functions, which breaks things when `-Wl,-Bsymbolic-functions` is also specified.

So apparently the real problem is that `clang` in the `nixos-22.05` branch uses wrong library paths for the GCC libraries.

**Comment by mweinelt** [comment]
I'm currently building with the following patch, thanks to trofi.

```
diff --git a/pkgs/applications/networking/browsers/firefox/common.nix b/pkgs/applications/networking/browsers/firefox/common.nix
index aadbb32d6e02..7a00670ebc44 100644
--- a/pkgs/applications/networking/browsers/firefox/common.nix
+++ b/pkgs/applications/networking/browsers/firefox/common.nix
@@ -497,6 +497,8 @@ buildStdenv.mkDerivation ({
 
   preBuild = ''
     cd mozobj
+  '' + lib.optionalString (stdenv.cc.isGNU) ''
+    export NIX_LDFLAGS="-L${stdenv.cc.cc.lib}/lib"
   '' + lib.optionalString (lib.versionAtLeast version "120") ''
     # https://bugzilla.mozilla.org/show_bug.cgi?id=1864083
     export NIX_CFLAGS_COMPILE="$NIX_CFLAGS_COMPILE $(pkg-config dbus-1 --cflags)"

**Comment by mweinelt** [comment]
This will most likely be resolved through https://github.com/NixOS/nixpkgs/pull/267737, but I'm still testing a solution that would be more localized to firefox.

**Comment by mweinelt** [comment]
The fix is part of the latest release-23.05 evaluation. You can follow its progress here: https://nixpk.gs/pr-tracker.html?pr=269680
